src/circuit/gadget/ecc/chip.rs

```diff
@@ -0,0 +1,423 @@
+            q_mul_hi: meta.selector(),
+            q_mul_lo: meta.selector(),
+            q_mul_decompose_var: meta.selector(),
+            q_mul_complete: meta.selector(),
+            q_mul_fixed: meta.selector(),
+            q_mul_fixed_short: meta.selector(),
+            q_point: meta.selector(),
+            q_scalar_fixed: meta.selector(),
+            q_scalar_fixed_short: meta.selector(),
+            perm: Permutation::new(
+                meta,
+                &[
+                    advices[0].into(),
+                    advices[1].into(),
+                    advices[2].into(),
+                    advices[3].into(),
+                    advices[4].into(),
+                    advices[6].into(),
+                    advices[7].into(),
+                    advices[8].into(),
+                    advices[9].into(),
+                ],
+            ),
+        };
+
+        // Create witness point gate
+        {
+            let config: witness_point::Config = (&config).into();
+            config.create_gate::<C>(meta);
+        }
+
+        // Create witness scalar_fixed gates (both full-width and short)
+        {
+            let config: witness_scalar_fixed::Config = (&config).into();
+            config.create_gate::<C>(meta);
+        }
+
+        // Create incomplete point addition gate
+        {
+            let config: add_incomplete::Config = (&config).into();
+            config.create_gate(meta);
+        }
+
+        // Create complete point addition gate
+        {
+            let add_config: add::Config = (&config).into();
+            add_config.create_gate(meta);
+        }
+
+        // Create fixed-base scalar mul gates
+        {
+            let mul_fixed_config: mul_fixed::Config = (&config).into();
+            mul_fixed_config.create_gate::<C>(meta);
+        }
+
+        // Create variable-base scalar mul gates
+        {
+            let mul_config: mul::Config<C> = (&config).into();
+            mul_config.create_gate(meta);
+        }
+
+        config
+    }
+
+    pub fn load() -> <Self as Chip<C::Base>>::Loaded {
+        let commit_ivk_r = load::commit_ivk_r();
+        let note_commit_r = load::note_commit_r();
+        let nullifier_k = load::nullifier_k();
+        let value_commit_r = load::value_commit_r();
+        let value_commit_v = load::value_commit_v();
+
+        EccLoaded {
+            commit_ivk_r,
+            note_commit_r,
+            nullifier_k,
+            value_commit_r,
+            value_commit_v,
+        }
+    }
+}
+
+/// A full-width scalar used for variable-base scalar multiplication.
+/// This is decomposed in chunks of `window_width` bits in little-endian order.
+/// For example, if `window_width` = 3, we will have [k_0, k_1, ..., k_n]
+/// where `scalar = k_0 + k_1 * (2^3) + ... + k_n * (2^3)^n`.
+#[derive(Clone, Debug)]
+pub struct EccScalarFixed<C: CurveAffine> {
+    value: Option<C::Scalar>,
+    k_bits: Vec<CellValue<C::Base>>,
```

Comment:
```suggestion
    k_limbs: Vec<CellValue<C::Base>>,
```

Reply:
I've renamed `k_bits` -> `windows` for fixed-base scalar mul.